NullPointerException when combining field and script in aggregation

Using a script in the percentile aggregation in Kibana 4, this results in an Elasticsearch NullPointerException. [Example](https://www.dropbox.com/s/op3kzha6qwjlbpe/Screenshot%202015-10-23%2014.08.17.png?dl=0). This seems to be caused by Kibana including both the field and script property. It is not possible to exclude the field property from Kibana.

For the sake of simplicity I narrowed this down to a plain simple ES query. Elasticsearch 1.7.3 on a bare CentOS 7.1 box installed using the yum repo. No changes have been made to the ES config.

Be sure to have one document in an index with a numerical field.

    $ curl -XDELETE localhost:9200/test
    
    $ curl -XPUT localhost:9200/test/test/1?pretty=true -d '{
        "metric": 100
    }'
    
    {
      "_index" : "test",
      "_type" : "test",
      "_id" : "1",
      "_version" : 1,
      "created" : true
    }

Mapping has become a long value, which is fine for percentile agg

    $ curl -XGET localhost:9200/test/_mapping?pretty=true
    
    {
      "test" : {
        "mappings" : {
          "test" : {
            "properties" : {
              "metric" : {
                "type" : "long"
              }
            }
          }
        }
      }
    }

Running a query with both the field and script properties supplied. Results in a NullPointerException. I expected that the script would have precedence.

    $ curl -XPOST localhost:9200/test/test/_search?pretty=true -d '{
      "size": 0,
      "query": {
        "match_all": {}
      },
      "aggs": {
        "some-agg": {
          "percentiles": {
            "field": "metric",
            "lang": "expression",
            "script": "1"
          }
        }
      }
    }'
    
    {
      "took" : 4,
      "timed_out" : false,
      "_shards" : {
        "total" : 5,
        "successful" : 4,
        "failed" : 1,
        "failures" : [ {
          "index" : "test",
          "shard" : 2,
          "status" : 500,
          "reason" : "QueryPhaseExecutionException[[test][2]: query[ConstantScore(cache(_type:test))],from[0],size[0]: Query Failed [Failed to execute main query]]; nested: NullPointerException; "
        } ]
      },
      "hits" : {
        "total" : 0,
        "max_score" : 0.0,
        "hits" : [ ]
      },
      "aggregations" : {
        "some-agg" : {
          "values" : {
            "1.0" : "NaN",
            "5.0" : "NaN",
            "25.0" : "NaN",
            "50.0" : "NaN",
            "75.0" : "NaN",
            "95.0" : "NaN",
            "99.0" : "NaN"
          }
        }
      }
    }

PS; this is what is included in the Elasticsearch log file:
```
[2015-10-23 16:29:19,142][DEBUG][action.search.type       ] [Pip the Troll] [test][2], node[YWWDb9UsQEGefx8R1f_Urw], [P], s[STARTED]: Failed to execute [org.elasticsearch.action.search.SearchRequest@35c01a59] lastShard [true]
org.elasticsearch.search.query.QueryPhaseExecutionException: [test][2]: query[ConstantScore(cache(_type:test))],from[0],size[0]: Query Failed [Failed to execute main query]
	at org.elasticsearch.search.query.QueryPhase.execute(QueryPhase.java:163)
	at org.elasticsearch.search.SearchService.loadOrExecuteQueryPhase(SearchService.java:301)
	at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:312)
	at org.elasticsearch.search.action.SearchServiceTransportAction$5.call(SearchServiceTransportAction.java:231)
	at org.elasticsearch.search.action.SearchServiceTransportAction$5.call(SearchServiceTransportAction.java:228)
	at org.elasticsearch.search.action.SearchServiceTransportAction$23.run(SearchServiceTransportAction.java:559)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.NullPointerException
	at org.elasticsearch.script.expression.ExpressionScript.setNextVar(ExpressionScript.java:109)
	at org.elasticsearch.search.aggregations.support.ValuesSource$Numeric$WithScript$DoubleValues.setDocument(ValuesSource.java:513)
	at org.elasticsearch.search.aggregations.metrics.percentiles.AbstractPercentilesAggregator.collect(AbstractPercentilesAggregator.java:83)
	at org.elasticsearch.search.aggregations.AggregationPhase$AggregationsCollector.collect(AggregationPhase.java:161)
	at org.elasticsearch.common.lucene.MultiCollector.collect(MultiCollector.java:60)
	at org.apache.lucene.search.Weight$DefaultBulkScorer.scoreAll(Weight.java:193)
	at org.apache.lucene.search.Weight$DefaultBulkScorer.score(Weight.java:163)
	at org.apache.lucene.search.BulkScorer.score(BulkScorer.java:35)
	at org.apache.lucene.search.IndexSearcher.search(IndexSearcher.java:621)
	at org.elasticsearch.search.internal.ContextIndexSearcher.search(ContextIndexSearcher.java:191)
	at org.apache.lucene.search.IndexSearcher.search(IndexSearcher.java:309)
	at org.elasticsearch.search.query.QueryPhase.execute(QueryPhase.java:117)
	... 8 more
```

--- a/modules/lang-expression/src/main/java/org/elasticsearch/script/expression/ExpressionSearchScript.java
+++ b/modules/lang-expression/src/main/java/org/elasticsearch/script/expression/ExpressionSearchScript.java
@@ -112,14 +112,16 @@ class ExpressionSearchScript implements SearchScript {
 
             @Override
             public void setNextVar(String name, Object value) {
-                assert(specialValue != null);
                 // this should only be used for the special "_value" variable used in aggregations
                 assert(name.equals("_value"));
 
-                if (value instanceof Number) {
-                    specialValue.setValue(((Number)value).doubleValue());
-                } else {
-                    throw new ScriptException("Cannot use expression with text variable using " + compiledScript);
+                // _value isn't used in script if specialValue == null
+                if (specialValue != null) {
+                    if (value instanceof Number) {
+                        specialValue.setValue(((Number)value).doubleValue());
+                    } else {
+                        throw new ScriptException("Cannot use expression with text variable using " + compiledScript);
+                    }
                 }
             }
         };

--- a/modules/lang-expression/src/test/java/org/elasticsearch/script/expression/MoreExpressionTests.java
+++ b/modules/lang-expression/src/test/java/org/elasticsearch/script/expression/MoreExpressionTests.java
@@ -383,7 +383,11 @@ public class MoreExpressionTests extends ESIntegTestCase {
                                 .script(new Script("_value * 3", ScriptType.INLINE, ExpressionScriptEngineService.NAME, null)))
                 .addAggregation(
                         AggregationBuilders.stats("double_agg").field("y")
-                                .script(new Script("_value - 1.1", ScriptType.INLINE, ExpressionScriptEngineService.NAME, null)));
+                                .script(new Script("_value - 1.1", ScriptType.INLINE, ExpressionScriptEngineService.NAME, null)))
+                .addAggregation(
+                        AggregationBuilders.stats("const_agg").field("x")
+                                .script(new Script("3.0", ScriptType.INLINE, ExpressionScriptEngineService.NAME, null))
+                );
 
         SearchResponse rsp = req.get();
         assertEquals(3, rsp.getHits().getTotalHits());
@@ -395,6 +399,11 @@ public class MoreExpressionTests extends ESIntegTestCase {
         stats = rsp.getAggregations().get("double_agg");
         assertEquals(0.7, stats.getMax(), 0.0001);
         assertEquals(0.1, stats.getMin(), 0.0001);
+
+        stats = rsp.getAggregations().get("const_agg");
+        assertThat(stats.getMax(), equalTo(3.0));
+        assertThat(stats.getMin(), equalTo(3.0));
+        assertThat(stats.getAvg(), equalTo(3.0));
     }
 
     public void testStringSpecialValueVariable() throws Exception {

